include/picongpu/plugins/openPMD/openPMDWriter.hpp

```diff
@@ -0,0 +1,1446 @@
+#include "picongpu/plugins/misc/SpeciesFilter.hpp"
+#include "picongpu/plugins/misc/misc.hpp"
+#include "picongpu/plugins/multi/IHelp.hpp"
+#include "picongpu/plugins/multi/Option.hpp"
+#include "picongpu/plugins/openPMD/openPMDWriter.def"
+#include "picongpu/simulation/control/MovingWindow.hpp"
+#include "picongpu/simulation_defines.hpp"
+#include "picongpu/traits/IsFieldDomainBound.hpp"
+
+#include <pmacc/assert.hpp>
+#include <pmacc/dataManagement/DataConnector.hpp>
+#include <pmacc/dimensions/GridLayout.hpp>
+#include <pmacc/mappings/simulation/GridController.hpp>
+#include <pmacc/mappings/simulation/SubGrid.hpp>
+#include <pmacc/math/Vector.hpp>
+#include <pmacc/particles/IdProvider.def>
+#include <pmacc/particles/frame_types.hpp>
+#include <pmacc/particles/operations/CountParticles.hpp>
+#include <pmacc/pluginSystem/PluginConnector.hpp>
+#include <pmacc/static_assert.hpp>
+#if( PMACC_CUDA_ENABLED == 1 )
+#    include <pmacc/particles/memory/buffers/MallocMCBuffer.hpp>
+#endif
+#include "picongpu/plugins/misc/SpeciesFilter.hpp"
+#include "picongpu/plugins/openPMD/NDScalars.hpp"
+#include "picongpu/plugins/openPMD/WriteMeta.hpp"
+#include "picongpu/plugins/openPMD/WriteSpecies.hpp"
+#include "picongpu/plugins/openPMD/restart/LoadSpecies.hpp"
+#include "picongpu/plugins/openPMD/restart/RestartFieldLoader.hpp"
+#include "picongpu/plugins/output/IIOBackend.hpp"
+
+#include <pmacc/traits/Limits.hpp>
+
+#include <boost/filesystem.hpp>
+#include <boost/mpl/at.hpp>
+#include <boost/mpl/begin_end.hpp>
+#include <boost/mpl/find.hpp>
+#include <boost/mpl/pair.hpp>
+#include <boost/mpl/size.hpp>
+#include <boost/mpl/vector.hpp>
+#include <boost/type_traits.hpp>
+#include <boost/type_traits/is_same.hpp>
+
+#include <openPMD/openPMD.hpp>
+
+#if !defined( _WIN32 )
+#    include <unistd.h>
+#endif
+
+#include <algorithm>
+#include <cstdint>
+#include <list>
+#include <pthread.h>
+#include <sstream>
+#include <string>
+#include <vector>
+
+
+namespace picongpu
+{
+namespace openPMD
+{
+    using namespace pmacc;
+
+
+    namespace po = boost::program_options;
+
+    template< unsigned DIM >
+    ::openPMD::RecordComponent &
+    ThreadParams::initDataset(
+        ::openPMD::RecordComponent & recordComponent,
+        ::openPMD::Datatype datatype,
+        pmacc::math::UInt64< DIM > const & globalDimensions,
+        bool compression,
+        std::string const & compressionMethod )
+    {
+        std::vector< uint64_t > v = asStandardVector( globalDimensions );
+        ::openPMD::Dataset dataset{ datatype, std::move( v ) };
+        if( compression && compressionMethod != "none" )
+        {
+            dataset.compression = compressionMethod;
+        }
+        recordComponent.resetDataset( std::move( dataset ) );
+        return recordComponent;
+    }
+
+
+    template< typename Vec, typename Ret >
+    Ret
+    asStandardVector( Vec const & v )
+    {
+        using Vec_T = typename std::remove_reference< Vec >::type;
```

Comment:
Is this trait needed. `Vec` should never be a reference.

Reply:
In some instantiations `Vec` is `std::vector<...>&&`, so it is needed to remove that.